### PR TITLE
Add constraint support

### LINKED ISSUE #69: `@check` convenience syntax for record/field-level value constraints
Split out from #20. With Legolas v0.5 (ref #54), we have a much nicer disciplined stance on validation-by-construction, so the path forward for convenient record-level check constraints is much simpler. 

We should define a simple helper `@check` that can be used in the following manner:

```jl
@version FooV1 begin
    x::String # unchecked
    @check(y::Int, y > 0)
    @check(z::Any, a_predicate(z) && another_predicate(z, x))
    @check(k::Int = clamp(k, 1, 10), 
           some_other_predicate(k))
    @check begin
        # record-level predicate(s)
    end
end
```

where the field-level `@check` expansion goes from statements like

```jl
@version FooV1 begin
    ⋮
    @check(f::T = $rhs, $predicate)
    ⋮
end
```

...to statements like:

```jl
f::T = $rhs
$predicate || throw(CheckConstraintError(:FooV1, :f, f, $stringified_predicate))
```

...where `CheckConstraintError` might look like:

```jl
struct CheckConstraintError <: Exception
    record_name::Symbol
    field_name::Symbol
    predicate::String
    received::Any
end

function Base.showerror(io::IO, e::CheckConstraintError)
    print(io, """
              Value received for $(e.field_name) of $(e.record_name): $(e.received)
              Violates `@check` constraint: $(e.predicate)
              """)
end
```

### PR DESCRIPTION
**Pull request by omus**
Adds basic constraint support for Legolas schemas so you can define constraints on the relationships between fields. Previously, this could be worked around by adding such logic into a field:

```julia
julia> using Legolas: @schema, @version

julia> @schema "old.constraint" OldConstraint

julia> @version OldConstraintV1 begin
           a
           b = begin
               @assert a == b
               b
           end
       end

julia> OldConstraintV1(; a=1, b=2)
ERROR: ArgumentError: Invalid value set for field `b` (2)
Stacktrace:
 [1] OldConstraintV1(; a::Int64, b::Int64)
   @ Main ~/.julia/dev/Legolas/src/schemas.jl:545
 [2] top-level scope
   @ REPL[4]:1

caused by: AssertionError: a == b
Stacktrace:
 [1] macro expansion
   @ ./REPL[3]:4 [inlined]
 [2] OldConstraintV1(; a::Int64, b::Int64)
   @ Main ~/.julia/dev/Legolas/src/schemas.jl:580
 [3] top-level scope
   @ REPL[4]:1
```

Unfortunately, the old approach would produce an extra `ArgumentError` which could create confusion. Additionally, when utilizing more complicated constructor logic for `b` these kinds of constraints can be less clear than declaring them separately like so:

```julia
julia> using Legolas: @schema, @version

julia> @schema "new.constraint" NewConstraint

julia> @version NewConstraintV1 begin
           a
           b
           @check a == b
       end

julia> NewConstraintV1(; a=1, b=2)
ERROR: Legolas.CheckConstraintError: a == b
Stacktrace:
 [1] macro expansion
   @ ./REPL[3]:4 [inlined]
 [2] NewConstraintV1(; a::Int64, b::Int64)
   @ Main ~/.julia/dev/Legolas/src/schemas.jl:619
 [3] top-level scope
   @ REPL[4]:1
```

Closes #69

### CODE REVIEW

**Comment by omus** [comment]
GitHub actions is currently having an outage

**Comment by jrevels** [comment]
xref https://github.com/beacon-biosignals/Legolas.jl/issues/69

I'd like to use `@check` for this instead of `@assert`; want to avoid teaching users that it's literally the same as Julia's built-in`@assert`, as we might add more sugar/behaviors later to this (e.g. https://github.com/beacon-biosignals/Legolas.jl/issues/20)

**Comment by omus** [comment]
> I'd like to use `@check` for this instead of `@assert`; want to avoid teaching users that it's literally the same as Julia's built-in`@assert`, as we might add more sugar/behaviors later to this (e.g. #20)

Good call. I started off with `@assert` just to prove this was a viable approach. I will update this PR to use `@check`

File: test/runtests.jl
Comment (kleinschmidt) [inline test/runtests.jl:876]: ??

File: src/schemas.jl
Comment (kleinschmidt) [inline src/schemas.jl:796]: this is pretty opaque to me, add a comment explaining why this is necessary?

Comment (kleinschmidt) [reply src/schemas.jl:796]: we don't want to just strip out the line number nodes right, since that's what will give the nice error w/ teh correct definition location when the check fails, right?

Comment (kleinschmidt) [reply src/schemas.jl:796]: ah wait I think I see: we're creating a new `begin...end` with the line number node in the macrocall manually spliced in (if there is one).

**Review by kleinschmidt — COMMENTED** [review]
I'm generally good with this.  I think a bit of explaining about the line number node manipulations would be helpful for future maintainers.

How reliable is the line-number manipulation across julia versions?  I have a vague sense that it might uh not be the most reliable thing but I have nothing to back that up...

File: test/runtests.jl
Comment (omus) [reply test/runtests.jl:876]: Ensures that if `ConstraintV1(; a=1, b=missing)` does not thrown an exception that a test failure is reported.

Comment (kleinschmidt) [reply test/runtests.jl:876]: ah okay, could use a comment too.

File: src/schemas.jl
Comment (omus) [reply src/schemas.jl:795]: I've updated the comment to something which I think is clearer

**Review by kleinschmidt — APPROVED** [review]

**Comment by omus** [comment]
I'll be waiting until https://github.com/beacon-biosignals/Legolas.jl/pull/123 is merged before merging this.

**Comment by omus** [comment]
#123 has been merged. I need to update the version here and finish some downstream testing I started.

**Pull request merged**
